### ISSUE TITLE
Autogluon does not gracefully handle sparse pandas dataframes

### ISSUE THREAD

**Issue by pushpendre** [awslabs/autogluon#567]
Consider the following MWE where I created a sparse 10,000 × 1,500 matrix and, I made the first column the label. The entire dataset takes less only a few MB in memory. The `task.fit` step encounters two major error. 

```python
import autogluon as ag
import pandas as pd
from scipy import sparse
from autogluon import TabularPrediction as task
n = int(1e4)
m = sparse.eye(n, format='csc')[:, :1500]
m[::2, 0] = 1 # make it a balanced binary prediction problem.
columns=['label', *range(1499)]

dataset = task.Dataset(df=pd.DataFrame.sparse.from_spmatrix(m, columns=columns)) 
predictor = task.fit(train_data=dataset, 
                     label='label', 
                     output_directory='/tmp/tmp')
```

**First** the main issue is that autogluon outputs the following error 1499 times, presumably once for each column!

```
ERROR:autogluon.utils.tabular.features.utils:Warning: dtype Sparse[float64, 0.0] is not recognized as a valid dtype by numpy! AutoGluon may incorrectly handle this feature...
Traceback (most recent call last):
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/features/utils.py", line 16, in get_type_family
    elif np.issubdtype(dtype, np.integer):
  File "/home/prastogi/conda/lib/python3.7/site-packages/numpy/core/numerictypes.py", line 393, in issubdtype
    arg1 = dtype(arg1).type
TypeError: data type not understood
ERROR:autogluon.utils.tabular.features.utils:data type not understood
Traceback (most recent call last):
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/features/utils.py", line 16, in get_type_family
    elif np.issubdtype(dtype, np.integer):
  File "/home/prastogi/conda/lib/python3.7/site-packages/numpy/core/numerictypes.py", line 393, in issubdtype
    arg1 = dtype(arg1).type
TypeError: data type not understood
```

**Second** it outputs the following error with the NeuralNetclassifier.

```
INFO:root:Fitting model: NeuralNetClassifier ...
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 255, in train_and_save
    model = self.train_single(X_train, y_train, X_val, y_val, model, kfolds=kfolds, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, level=level, time_limit=time_limit)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 240, in train_single
    model.fit(X_train=X_train, y_train=y_train, X_val=X_val, y_val=y_val, time_limit=time_limit, **model_fit_kwargs)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/abstract/abstract_model.py", line 264, in fit
    self._fit(**kwargs)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py", line 206, in _fit
    self.train_net(train_dataset=train_dataset, params=params, val_dataset=val_dataset, initialize=True, setup_trainer=True, time_limit=time_limit, reporter=reporter)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py", line 328, in train_net
    output = self.model(data_batch)
  File "/home/prastogi/conda/lib/python3.7/site-packages/mxnet/gluon/block.py", line 693, in __call__
    out = self.forward(*args)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/tabular_nn/embednet.py", line 210, in forward
    return self.output_block(input_activations)
ERROR:autogluon.utils.tabular.ml.trainer.abstract_trainer:Warning: Exception caused NeuralNetClassifier to fail during training... Skipping this model.
Traceback (most recent call last):
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 255, in train_and_save
    model = self.train_single(X_train, y_train, X_val, y_val, model, kfolds=kfolds, k_fold_start=k_fold_start, k_fold_end=k_fold_end, n_repeats=n_repeats, n_repeat_start=n_repeat_start, level=level, time_limit=time_limit)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/trainer/abstract_trainer.py", line 240, in train_single
    model.fit(X_train=X_train, y_train=y_train, X_val=X_val, y_val=y_val, time_limit=time_limit, **model_fit_kwargs)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/abstract/abstract_model.py", line 264, in fit
    self._fit(**kwargs)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py", line 206, in _fit
    self.train_net(train_dataset=train_dataset, params=params, val_dataset=val_dataset, initialize=True, setup_trainer=True, time_limit=time_limit, reporter=reporter)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/tabular_nn/tabular_nn_model.py", line 328, in train_net
    output = self.model(data_batch)
  File "/home/prastogi/conda/lib/python3.7/site-packages/mxnet/gluon/block.py", line 693, in __call__
    out = self.forward(*args)
  File "/home/prastogi/conda/lib/python3.7/site-packages/autogluon/utils/tabular/ml/models/tabular_nn/embednet.py", line 210, in forward
    return self.output_block(input_activations)
UnboundLocalError: local variable 'input_activations' referenced before assignment
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:local variable 'input_activations' referenced before assignment
```

**However** the tree based classifiers and the KNN classifier models seem to work.

```
INFO:root:Fitting model: RandomForestClassifierGini ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	7.98s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.42s	 = Validation runtime

INFO:root:Fitting model: RandomForestClassifierEntr ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	7.54s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.25s	 = Validation runtime

INFO:root:Fitting model: ExtraTreesClassifierGini ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	13.05s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.45s	 = Validation runtime

INFO:root:Fitting model: ExtraTreesClassifierEntr ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	13.41s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.62s	 = Validation runtime

INFO:root:Fitting model: KNeighborsClassifierUnif ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.23s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.57s	 = Validation runtime

INFO:root:Fitting model: KNeighborsClassifierDist ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.21s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.48s	 = Validation runtime

INFO:root:Fitting model: LightGBMClassifier ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	2.13s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.68s	 = Validation runtime

INFO:root:Fitting model: CatboostClassifier ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.62s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.04s	 = Validation runtime

INFO:root:Fitting model: LightGBMClassifierCustom ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	1.67s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.65s	 = Validation runtime

INFO:root:Fitting model: weighted_ensemble_k0_l1 ...
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.5	 = Validation accuracy score
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.84s	 = Training runtime
INFO:autogluon.utils.tabular.ml.trainer.abstract_trainer:	0.0s	 = Validation runtime

INFO:autogluon.utils.tabular.ml.learner.default_learner:AutoGluon training complete, total runtime = 98.8s ...
```

It'll be great if the

**Comment by pushpendre** [comment]
a workaround is to configure autogluon as follows

```python
import autogluon.utils.tabular.features.utils
autogluon.utils.tabular.features.utils.logger.disabled = True
predictor = task.fit(..., excluded_model_types=['NN'])
```

**Comment by Innixma** [comment]
Thanks for creating an issue!

I'm genuinely surprised that AutoGluon didn't outright crash with those inputs. To be clear, AutoGluon does not support sparse input data at present, and I suspect the models aren't actually using the features you provided during training.

I do think that we can do a better job regarding the logging of errors in this situation, by more clearly indicating that the features are invalid because they are sparse.

Going a step further, we could attempt to automatically convert sparse features to dense features (with memory usage checks to avoid OOM).

Contributions are welcome!

**Comment by pushpendre** [comment]
I was actually going into the internals of auto-gluon yesterday to try and make this thing work, since I'll really prefer to use auto-gluon as a top-level orchestrator for lightgbm + timed Hyper Parameter Optimization. I.e. even if I don't end up using any thing other than lightgbm I think the automatic feature pre-processing and time-sensitive hyper-parameter optimization are good features.

I think the first issue comes mainly from

```
autogluon/utils/tabular/features/abstract_feature_generator.py
```

This class does not understand the `Sparse[float64, 0.0]` data type extension that pandas provides.

> To be clear, AutoGluon does not support sparse input data at present, and I suspect the models aren't actually using the features you provided during training.

oh that's too bad. I thought that gluon could do something smart-ish with the features, for example lightgbm itself can definitely handle sparse matrices. Other predictors should also be able to handle these values. I'll do some more testing before throwing in the towel.


> I do think that we can do a better job regarding the logging of errors in this situation, by more clearly indicating that the features are invalid because they are sparse.


> Going a step further, we could attempt to automatically convert sparse features to dense features (with memory usage checks to avoid OOM).

FWIW I'd vote against this. This is not really a *value-add* because the work-around of making the entire feature-set dense is so simple. At least if we were making only a mini-batch dense that will be a scalable solution.

**Comment by Innixma** [comment]
Regarding AutoGluon not supporting sparse data, this is primarily due to abstract_feature_generator which is likely just dropping the features because it doesn't recognize them. If we add functionality for it to recognize them (quite simple), they should pass through feature generation and be fed into the ML models.

The next step will be to update the .fit() functions of the models to handle sparse input without error. This would be where we would likely automatically convert to dense if the model doesn't have support for sparse (or just drop the feature entirely from the model). If a model does support sparse (XGBoost, LightGBM, Linear model, etc.), then we would add logic to properly handle them within the model class.

This can be added through several PR's so long as models which don't yet have sparse support still are able to train on the non-sparse features (or convert sparse to dense), therefore avoiding crashes.

**Comment by pushpendre** [comment]
@Innixma here's a minimal patch that I wrote to appease the `abstract_feature_generator` and the `auto_ml_feature_generator.py`.
https://gist.github.com/pushpendre/a17ce02e8bc2ee293f93360495070954

With these changes my MWE passes

```bash
Feature Generator processed 10000 data points with 1499 features
Original Features (raw dtypes):
        Sparse[float64, 0.0] features: 1499
Original Features (inferred dtypes):
        float features: 1499
Generated Features (special dtypes):
Final Features (raw dtypes):
        float features: 1499
Final Features:
        float features: 1499
        Data preprocessing and feature engineering runtime = 97.52s ...
```

Before putting in more time I wanted to ask a couple of clarification questions 

1. Is the approach in the patch reasonable? 
1. Is there a flow-chart / diagram of what the `feature_generator` is trying to do? It's actually pretty hard to figure out what exactly all the feature processing code is supposed to do, and how to change it without breaking anything else.
2. There are lots of intermediate dicts that are created `self.features_init, self.feature_type_family, self.features_to_remove` but they are not always used later on when they should be. E.g. in many places the `X.dtypes` is directly used. This was confusing to me.
3. Could you point me to some test-cases used to check categorical/text/raw/date feature preprocessing? I could then make a sparse version of those and make sure that the tests pass.
4. What is the expected "state" after "feature_generation". I.e. if I as a user have already implemented things like normalization, one-hot-encoding, in a sklearn pipeline, and I want to give the output of that pipeline to auto-gluon, what flag do I need to tell auto-gluon to skip preprocessing, and what format does it need to be in?

Thanks

**Comment by Innixma** [comment]
@pushpendre Great work regarding your changes. Your comments and questions are much appreciated.

I have made an issue (#579) regarding refactoring the feature generator so my plans are more widely visible. The feature generator is the oldest and least maintained part of the code at present, and I want to change that.

Regarding your questions:

1. Yes it appears reasonable (With some comments I will make later on).
2. Not at present, and I agree it is hard to understand. This should be improved during the refactor effort.
3. Will go variable by variable: 
    - `self.features_init` is not necessary, and I will remove it, good catch!
    - `self.feature_type_family` is necessary. Do note that X.dtypes != the values in feature_type_family. X.dtypes can contain `int64`, `int32`, `int16`, `int8`, but feature_type_family will consider all of these `int`. We use X.dtypes when we care about the true exact type of the feature. Looking closer at the code, I do see it used unnecessarily once in `AutoMLFeatureGenerator`, but in `AbstractFeatureGenerator` it is used to store in `self.features_init_types`, so that we can ensure the dtypes at inference time before further processing are **_exactly_** the same as during training.
    - `self.features_to_remove` is likely not necessary. I think using `self.features_init_to_keep` in its place is better.
4. Unfortunately there are no dedicated tests for the feature generator at present, and this is a top priority for the refactor effort. Previously we only dealt with very simple standard feature types you could get from loading a raw csv, so our end-to-end benchmarks producing strong results without crashes was enough evidence that the feature generator was working properly. This of course changes when we start introducing more complex features such as sparse features where it becomes difficult to test all of the possible input types in a typical benchmark.
5. This level of customization is not available at present, but is part of what I want to accomplish in the refactor. The key to enabling this will be the [FeatureTypesMetadata](https://github.com/awslabs/autogluon/blob/master/autogluon/utils/tabular/features/feature_types_metadata.py#L7) object, which a user will be able to construct and pass in as input to the feature generator. This will allow them to indicate which features to skip generation for, enforce the type of certain features, etc.

Regarding your specific code changes:

1. I'd recommend adding 'sparse' as a `special dtype` that contains the list of all sparse features, similar to how `text_as_category` contains categorical features that were originally text. This should help downstream models determine how to process these sparse features, and will allow us to add generic logic to drop all the sparse features for certain models / convert to dense.
2. I'm unsure if we should have sparse floats be considered 'float' as their type family instead of 'sparse_float', but we can figure that out later based on whichever is easier from the model usage side. By default, I think many of the models will crash if they try to treat sparse floats as dense floats.
3. `self.features_init_types` should stay as is, we want the exact types, not the type family.
4. `object_column_set` is interesting. It might make sense to have all of our feature dictionaries be sets.
5. L49: change to `if self.features_init_types[column].name == 'object':`. Perhaps it might be useful to also have this representation for feature_types_family (feature name -> dtype_family mapping), and use that variable instead. That way any of 'string', 'str', and 'object' dtypes would get the proper treatment.

Finally, if you are interested in working more closely with the team and participating in discussions, I'd be happy to invite you to our AutoGluon contributor slack channel. If you are interested, just send me an email at innixma@gmail.com or autogluon-dev@amazon.com

Thanks!

**Comment by pushpendre** [comment]
I went ahead and did some work to use Pandas SparseDataframe with lightgbm. It turns out that lightgbm does not support the extension dtype `Sparse[]` that is defined in pandas. It accepted this datatype but then started converting to dense matrix. But thankfully lightgbm does support -- and I have tested this myself -- CSR/CSC sparse matrices. So just focusing on a case where the `train_data` is not a pandas dataframe but rather a sparse matrix and 

```
predictor = task.fit(train_data=dataset,  # an actual np.array
                     label=label, #actual np.array
                     output_directory='/tmp/tmp')
```

might make life easier.

**Comment by pushpendre** [comment]
Focusing on a case where the input is an array, can also help in adding a flag where the preprocessing can be skipped. If the user is passing in an array of `float` values then the user may have already done all the preprocessing themselves. The  user should be allowed to bypass feature pre-processing at this point. A benefit is that this way auto gluon will not have to deal with the sparse-feature pre-processing that may be very messy.

**Comment by Innixma** [comment]
> Focusing on a case where the input is an array, can also help in adding a flag where the preprocessing can be skipped. If the user is passing in an array of `float` values then the user may have already done all the preprocessing themselves. The user should be allowed to bypass feature pre-processing at this point. A benefit is that this way auto gluon will not have to deal with the sparse-feature pre-processing that may be very messy.

Definitely agree. I plan to eventually add this option after the feature generator refactor. This is actually a way to speed-up the inference time on bagged ensembles since we wouldn't have to repeat the preprocessing 10 times, only once, so long as the preprocessing is not stateful in regards to the training rows used to train the model. In this sense I think it will be best to have 2-stage preprocessing in models: `preprocess_stateless` followed by `preprocess_stateful`. Stateful cannot be skipped (or at least shouldn't be), but stateless can with an argument flag.

**Comment by Innixma** [comment]
> I went ahead and did some work to use Pandas SparseDataframe with lightgbm. It turns out that lightgbm does not support the extension dtype `Sparse[]` that is defined in pandas. It accepted this datatype but then started converting to dense matrix. But thankfully lightgbm does support -- and I have tested this myself -- CSR/CSC sparse matrices. So just focusing on a case where the `train_data` is not a pandas dataframe but rather a sparse matrix and
> 
> ```
> predictor = task.fit(train_data=dataset,  # an actual np.array
>                      label=label, #actual np.array
>                      output_directory='/tmp/tmp')
> ```
> 
> might make life easier.

Great! Feel free to have a new class inherit from LGBM as a prototype POC to do whatever you need to get it working with your sparse input. One thing to note is that you will have to specify the categorical features in this instance to LightGBM. In your implementation, you can assume that the model magically knows which features are categorical even if it isn't possible to tell from X_train alone.

I'd recommend using https://github.com/awslabs/autogluon/blob/master/examples/tabular/example_custom_model_tabular.py as a template, so you don't have to worry about making task.fit accept CSR/CSC matrices, as this will be significant work due to how we use pandas indices frequently which numpy ndarray's do not have. It will also be useful to have a DataFrame -> Sparse CSR/CSC matrix conversion function if you wish to write it.

**Comment by albert-ying** [comment]
Hi, I'm wondering if there is any update on this? Is it possible to input a sparse matrix to Autogluon now?

**Comment by Innixma** [comment]
@albert-ying Sparse matrices are not yet supported, however a great deal of work has been done to make adding this functionality easier via a total refactor of feature preprocessing in AutoGluon.

If anyone wants to try implementing this feature, first I'd recommend taking a close look at this example of building a custom feature generator:

https://github.com/awslabs/autogluon/blob/master/examples/tabular/example_custom_feature_generator.py

From this script, you'll be able to test any new functionality you add and pass sparse inputs to it to see if it is able to handle the sparse data without crashing.
